File: tests/example-program-gen/exp/accounts/State2.ts
import { PublicKey, Connection } from "@solana/web3.js"
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface State2Fields {
  vecOfOption: Array<BN | null>
}

export interface State2JSON {
  vecOfOption: Array<string | null>
}

export class State2 {
  readonly vecOfOption: Array<BN | null>

  static readonly discriminator = Buffer.from([
    106, 97, 255, 161, 250, 205, 185, 192,
  ])

  static readonly layout = borsh.struct([
    borsh.vec(borsh.option(borsh.u64()), "vecOfOption"),
  ])

  constructor(fields: State2Fields) {
    this.vecOfOption = fields.vecOfOption
  }

  static async fetch(
    c: Connection,
    address: PublicKey
  ): Promise<State2 | null> {
    const info = await c.getAccountInfo(address)

    if (info === null) {
      return null
    }
    if (!info.owner.equals(PROGRAM_ID)) {
      throw new Error("account doesn't belong to this program")
    }

    return this.decode(info.data)
  }

  static async fetchMultiple(
    c: Connection,
    addresses: PublicKey[]
  ): Promise<Array<State2 | null>> {
    const infos = await c.getMultipleAccountsInfo(addresses)

    return infos.map((info) => {
      if (info === null) {
        return null
      }
      if (!info.owner.equals(PROGRAM_ID)) {
        throw new Error("account doesn't belong to this program")
      }

      return this.decode(info.data)
    })
  }

  static decode(data: Buffer): State2 {
    if (!data.slice(0, 8).equals(State2.discriminator)) {
      throw new Error("invalid account discriminator")
    }

    const dec = State2.layout.decode(data.slice(8))

    return new State2({
      vecOfOption: dec.vecOfOption,
    })
  }

  toJSON(): State2JSON {
    return {
      vecOfOption: this.vecOfOption.map(
        (item) => (item && item.toString()) || null
      ),
    }
  }

  static fromJSON(obj: State2JSON): State2 {
    return new State2({
      vecOfOption: obj.vecOfOption.map(
        (item) => (item && new BN(item)) || null
      ),
    })
  }
}

File: tests/client-gen/example_program_gen/accounts/state2.py
import typing
from solana.publickey import PublicKey
from solana.rpc.async_api import AsyncClient
from solana.rpc.commitment import Commitment
import borsh_construct as borsh
from anchorpy.coder.accounts import ACCOUNT_DISCRIMINATOR_SIZE
from anchorpy.error import AccountInvalidDiscriminator
from ..program_id import PROGRAM_ID


class State2Fields(typing.TypedDict):
    vec_of_option: list[typing.Optional[int]]


class State2JSON(typing.TypedDict):
    vec_of_option: list[typing.Optional[int]]


class State2(object):
    discriminator = b"ja\xff\xa1\xfa\xcd\xb9\xc0"
    layout = borsh.CStruct("vec_of_option" / borsh.Vec(borsh.Option(borsh.U64)))

    def __init__(self, fields: State2Fields) -> None:
        self.vec_of_option = fields["vec_of_option"]

    @classmethod
    async def fetch(
        cls,
        conn: AsyncClient,
        address: PublicKey,
        commitment: typing.Optional[Commitment] = None,
    ) -> typing.Optional["State2"]:
        resp = await conn.get_account_info(address, commitment=commitment)
        info = resp["result"]["value"]
        if info is None:
            return None
        if info["owner"] != str(PROGRAM_ID):
            raise ValueError("Account does not belong to this program")
        return cls.decode(info["data"])

    @classmethod
    async def fetch_multiple(
        cls,
        conn: AsyncClient,
        addresses: list[typing.Union[PublicKey, str]],
        commitment: typing.Optional[Commitment] = None,
    ) -> typing.List[typing.Optional["State2"]]:
        resp = await conn.get_multiple_accounts(addresses, commitment=commitment)
        infos = resp["result"]["value"]
        res: typing.List[typing.Optional["State2"]] = []
        for info in infos:
            if info is None:
                res.append(None)
            if info["owner"] != str(PROGRAM_ID):
                raise ValueError("Account does not belong to this program")
            res.append(cls.decode(info["data"]))
        return res

    @classmethod
    def decode(cls, data: bytes) -> "State2":
        if data[:ACCOUNT_DISCRIMINATOR_SIZE] != cls.discriminator:
            raise AccountInvalidDiscriminator(
                "The discriminator for this account is invalid"
            )
        dec = State2.layout.decode(data[ACCOUNT_DISCRIMINATOR_SIZE:])
        return cls(
            {
                "vec_of_option": dec.vec_of_option,
            }
        )

    def to_json(self) -> State2JSON:
        return {
            "vec_of_option": self.vec_of_option,
        }

    @classmethod
    def from_json(cls, obj: State2JSON) -> "State2":
        return cls(
            {
                "vec_of_option": obj["vec_of_option"],
            }
        )
